File: java/src/cfalcione/cs303/lab03/MergeSort.java
package cfalcione.cs303.lab03;

import cfalcione.cs303.lab02.InsertionSort;
import cfalcione.cs303.shared.ArrayIterator;
import cfalcione.cs303.shared.SortingAlgorithm;

import java.util.Deque;
import java.util.LinkedList;

public class MergeSort <T extends Comparable<T>> extends SortingAlgorithm<T> {

    private int algorithmCutoff = 8;

    public MergeSort(){}

    public MergeSort(int algorithmCutoff) {
        this.algorithmCutoff = algorithmCutoff;
    }

    SortingAlgorithm<T> fallback = new InsertionSort<>();

    public String name() {
        return "Merge Sort";
    }


    public void sort(T[] array, int start, int end) {
        int length = end - start;
        if (length < 2) return;
        if (length < algorithmCutoff)
            fallback.sort(array, start, end);

        int mid = (start + end) / 2;
        sort(array, start, mid);
        sort(array, mid + 1, end);
        merge(array, start, mid, end);
    }

    public void merge(T[] array, int start, int mid, int end) {
        Deque<T> mergedList = buildMergedList(array, start, mid, end);
        for (int i = start; i <= end; i++) {
            array[i] = mergedList.pop();
        }
    }

    //Assumes that left and right are each sorted
    protected Deque<T> buildMergedList(T[] array, int start, int mid, int end) {
        ArrayIterator<T> left = new ArrayIterator<>(array, start, mid);
        ArrayIterator<T> right = new ArrayIterator<>(array, mid + 1, end);
        Deque<T> result = new LinkedList<>();

        while(left.hasNext() && right.hasNext()) {
            ArrayIterator<T> smaller = min(left, right);
            result.add(smaller.next());
        }

        while (left.hasNext()) {
            result.add(left.next());
        }

        while (right.hasNext()) {
            result.add(right.next());
        }

        return result;
    }

    private ArrayIterator<T> min(ArrayIterator<T> a, ArrayIterator<T> b) {
        int compare = a.peek().compareTo(b.peek());
        if (compare > 0) return b;
        return a;
    }

}

File: python/sorts/merge_sort.py
def make_merge_sort(inner_sort, threshold = 8):
    def sort(array, left, right):
        return merge_sort_util(inner_sort, threshold, array, left, right)
    return sort

def merge_sort_util(inner_sort, threshold, array, left, right):
    width = right - left
    if width <= threshold:
        inner_sort(array, left, right)
        return
    mid = (left + right) // 2
    merge_sort_util(inner_sort, threshold, array, left, mid)
    merge_sort_util(inner_sort, threshold, array, mid + 1, right)
    merge(array, left, mid, right)

def merge(array, left, mid, right):
    array[left:right+1] = merge_generator(array, left, mid, right)

def merge_generator(array, left, mid, right):
    left_iter = SubarrayIterator(array, left, mid)
    right_iter = SubarrayIterator(array, mid + 1, right)

    while left_iter and right_iter:
        yield next(min(left_iter, right_iter))
    
    yield from left_iter
    yield from right_iter



class SubarrayIterator():

    def __init__(self, array, start, end):
        self.copy = array[start:end+1]
        self.idx = 0
        self.end = len(self.copy) - 1
    
    def __bool__(self):
        return self.idx <= self.end
    
    def __lt__(self, other):
        return self.peek() < other.peek()
    
    def __iter__(self):
        return self
    
    def __next__(self):
        if not self:
            raise StopIteration
        result = self.peek()
        self.idx += 1
        return result
    
    def peek(self):
        return self.copy[self.idx]
    